tensorflow/contrib/kinesis/kernels/kinesis_dataset_ops.cc

```diff
@@ -0,0 +1,368 @@
+          TF_RETURN_IF_ERROR(SetupStreamsLocked(ctx->env()));
+        }
+        do {
+          Aws::Kinesis::Model::GetRecordsRequest request;
+          auto outcome = client_->GetRecords(
+              request.WithShardIterator(iterator_).WithLimit(1));
+          if (!outcome.IsSuccess()) {
+            string error =
+                strings::StrCat(outcome.GetError().GetExceptionName().c_str(),
+                                ": ", outcome.GetError().GetMessage().c_str());
+            return errors::Internal(error);
+          }
+          if (outcome.GetResult().GetRecords().size() == 0) {
+            // If return 0 record then nothing available at the moment.
+            if (dataset()->eof_) {
+              *end_of_sequence = true;
+              return Status::OK();
+            }
+            // Continue the loop after a period of time.
+            ctx->env()->SleepForMicroseconds(dataset()->interval_);
+            continue;
+          }
+          if (outcome.GetResult().GetRecords().size() != 1) {
+            return errors::Internal("invalid records number ",
+                                    outcome.GetResult().GetRecords().size(),
+                                    " returned");
+          }
+
+          iterator_ = outcome.GetResult().GetNextShardIterator();
+
+          StringPiece value(
+              (const char*)outcome.GetResult()
+                  .GetRecords()[0]
+                  .GetData()
+                  .GetUnderlyingData(),
+              outcome.GetResult().GetRecords()[0].GetData().GetLength());
+          Tensor value_tensor(ctx->allocator({}), DT_STRING, {});
+          value_tensor.scalar<std::string>()() = std::string(value);
+          out_tensors->emplace_back(std::move(value_tensor));
+
+          *end_of_sequence = false;
+          return Status::OK();
+        } while (true);
+      }
+
+     protected:
+      Status SaveInternal(IteratorStateWriter* writer) override {
+        return errors::Unimplemented("SaveInternal is currently not supported");
+      }
+
+      Status RestoreInternal(IteratorContext* ctx,
+                             IteratorStateReader* reader) override {
+        return errors::Unimplemented(
+            "RestoreInternal is currently not supported");
+      }
+
+     private:
+      // Sets up Kinesis streams to read from.
+      Status SetupStreamsLocked(Env* env) EXCLUSIVE_LOCKS_REQUIRED(mu_) {
+        AwsInitAPI();
+        client_.reset(
+            new Aws::Kinesis::KinesisClient(GetDefaultClientConfig()));
+
+        Aws::Kinesis::Model::DescribeStreamRequest request;
+        auto outcome = client_->DescribeStream(
+            request.WithStreamName(dataset()->stream_.c_str()));
+        if (!outcome.IsSuccess()) {
+          string error =
+              strings::StrCat(outcome.GetError().GetExceptionName().c_str(),
+                              ": ", outcome.GetError().GetMessage().c_str());
+          return errors::Internal(error);
```

Comment:
The `Internal` error type is usually reserved for violations of system invariants that weren't the user's fault. If this error can be triggered by invalid user input, use `InvalidArgument` or if it is more general that then, use `Unknown`.

Also, the error constructors in `errors::` accept multiple arguments that are implicitly concatenated, so the following is preferred:

```c++
  return errors::InvalidArgument(outcome.GetError().GetExceptionName().c_str(), ": ", 
                                 outcome.GetError().GetMessage().c_str());
```

Finally, are the `.c_str()` calls necessary? I think `Aws::String` is simply an alias for `std::string`, based on the API docs here:

https://sdk.amazonaws.com/cpp/api/0.14.3/namespace_aws.html#ac45eb95643c91062d4d82a28d453aae7

...so you can probably drop those calls.

Reply:
Thanks @mrry for the help. `c_str()` have been removed in most of the places (One place with `==` was kept. See other comments.)